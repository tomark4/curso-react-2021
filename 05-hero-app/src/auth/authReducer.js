import { types } from '../types/types'

const authReducer = (state = {}, action) => {
  // reducers para manejar el login
  switch(action.type){
    case types.login:
      return {...action.payload, logged: true}

    case types.logout:
      return { logged: false }

    default:
      return state
  }
}

export default authReducer;
