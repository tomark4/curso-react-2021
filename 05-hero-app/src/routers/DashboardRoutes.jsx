import Navbar from '../components/ui/Navbar';
import {
    Switch,
    Route,
    Redirect
} from 'react-router-dom';
import MarvelScreen from '../components/screens/MarvelScreen';
import HeroeScreen from '../components/screens/HeroeScreen';
import DcScreen from '../components/screens/DcScreen';
import SearchScreen from '../components/screens/SearchScreen';

const DashboardRoutes = () => {
    return (
        <>
            <Navbar />
            <div className="container mt-3">
                <Switch>
                    <Route exact path='/marvel' component={MarvelScreen} />
                    <Route exact path='/heroe/:id' component={HeroeScreen} />
                    <Route exact path='/dc' component={DcScreen} />
                    <Route exact path='/search' component={SearchScreen} />
                    <Redirect to="/marvel" />

                </Switch>
            </div>
        </>
    )
}

export default DashboardRoutes
