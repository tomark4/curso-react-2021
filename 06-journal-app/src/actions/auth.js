import { types } from '../types/types';
import {firebase, googleAuthProvider} from '../firebase';
import { startLoading, stopLoading } from './ui';
import Swal from 'sweetalert2';
import { noteLogout } from './notes';

// login with email and password
export const loginWithEmailPassword = (email, password) => (dispatch) => {
  dispatch( startLoading() );
  firebase.auth().signInWithEmailAndPassword(email, password)
  .then( ({user}) =>  {
    dispatch({
      type: types.login,
      payload: { uid: user.uid, name: user.displayName }
    });
    dispatch( stopLoading() );
  }).catch(err => {
    console.log(err);
    Swal.fire('Error',err.message, 'error')
    dispatch( stopLoading() );
  });
}

// login con google
export const loginWithGoogle = () => (dispatch) => {
  firebase.auth().signInWithPopup( googleAuthProvider )
  .then( ({user}) => {
    dispatch({
      type: types.login,
      payload: {uid: user.uid, name: user.displayName}
    })
  })
}

// register con email and password
export const registerWithEmailPassword = (email, password, name) => (dispatch) => {
  firebase.auth().createUserWithEmailAndPassword(email, password)
  .then( async ({user}) => {
    await user.updateProfile({displayName: name});
    dispatch({ type: types.login, payload: {uid: user.uid, name: user.displayName} });
  }).catch( err => console.log(err) );
}


export const startLogout = () => async (dispatch) => {
  try{
    await firebase.auth().signOut();
    dispatch( noteLogout() );
    dispatch( logout() );
  }catch(err){
    console.log(err)
  }
}

export const logout = () => ({
  type: types.logout
})
