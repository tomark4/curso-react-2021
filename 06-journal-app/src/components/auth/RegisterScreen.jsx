import {Link} from 'react-router-dom';

const RegisterScreen = () => {
    return (
        <>
            <h3 className='auth__title'>Registrar</h3>
            <form>
                <div>
                    <input
                    type="email"
                    placeholder="Email"
                    name="email"
                    className="auth__input"
                    autoComplete="off"/>
                </div>

                <div>
                    <input
                    type="text"
                    placeholder="Name"
                    name="name"
                    className="auth__input"
                    autoComplete="off"/>
                </div>

                <div>
                    <input
                    type="password"
                    placeholder="Password"
                    name="password"
                    className="auth__input"/>
                </div>

                <div>
                    <input
                    type="password"
                    placeholder="Password confirm"
                    name="password_confirm"
                    className="auth__input"/>
                </div>

                <div>
                    <button type="submit" className="btn btn-primary btn-block btn-lg">Registrar</button>
                </div>
            </form>
            <Link to="/auth/login" className="link">Â¿Ya tienes cuenta? Ingresar</Link>
        </>
    )
}

export default RegisterScreen
