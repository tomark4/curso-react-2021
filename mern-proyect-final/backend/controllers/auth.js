const { response } = require('express');
const Usuario = require("../models/Usuario");
const bcrypt = require("bcryptjs");
const { generateJWT } = require("../helpers/jwt");

const register = async (req, res = response) => {
  try{
    const { name, email, password } = req.body;

    // validar que el email no exista
    let usuario = await Usuario.findOne({email});

    if(usuario){
      return res.status(400).json({
        ok: false,
        message: 'usuario existe con ese correo'
      });
    }

    // crear si no existe
    usuario = new Usuario({name, email, password});

    // encriptar la constraseña
    const salt = bcrypt.genSaltSync();
    usuario.password = bcrypt.hashSync( password, salt);
    await usuario.save();

    const token = await generateJWT(usuario.id, usuario.name);

    return res.status(201).json({
      ok: true,
      uid: usuario.id,
      name: usuario.name,
      message: 'user created',
      token
    });

  } catch(err){
    console.log(err);
    return res.status(500).json({
      ok: false,
      message: 'error, contact support'
    });
  }

}


const login = async (req, res = response) => {
  try{
    const { email, password } = req.body;
    const usuario = await Usuario.findOne({email});

    if(!usuario){
      return res.status(400).json({
        ok: false,
        message: 'Usuario y contraseña no son correctos 1'
      });
    }

    // confirmar los password
    const validPassword = bcrypt.compareSync(password, usuario.password);

    if(!validPassword){
      return res.status(400).json({
        ok: false,
        message: 'Usuario y contraseña no son correctos 2'
      });
    }

    const token = await generateJWT(usuario.id, usuario.name);

    return res.status(200).json({
      ok: true,
      uid: usuario.id,
      name: usuario.name,
      message: 'login success',
      token
    });

  } catch(err){
    console.log(err);
    return res.status(500).json({
      ok: false,
      message: 'error contact support'
    });
  }
}


const refreshToken = async (req, res = response) => {
  try{

    const { uid, name } = req;
    const token = await generateJWT(uid, name);
    return res.json({ok: true, message:"refresh token", token});

  } catch( err ){
    console.log(err);
    return res.status(500).json({ok: false, message:"error"});
  }
}


module.exports = {
  register,
  login,
  refreshToken
};
