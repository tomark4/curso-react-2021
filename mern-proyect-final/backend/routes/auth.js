const express = require('express');
const router = express.Router();
const controller = require('../controllers/auth');
const {check} = require("express-validator");
const validateFields = require('../middlewares/validate-fields');
const { validateJWT } = require("../middlewares/validate-jwt");

// register user
router.post("/register",[
  check('name','Name is required').not().isEmpty(),
  check('email','Email is required').isEmail(),
  check('password','Password must have 6 characters').isLength({min: 6}),
  validateFields
], controller.register);

// login user
router.post("/login", [
  check('email','Email is required').isEmail(),
  check('password','Password must have 6 characters').isLength({min: 6}),
  validateFields
], controller.login);

// refresh token
router.get("/refresh-token", [validateJWT], controller.refreshToken);


module.exports = router;
