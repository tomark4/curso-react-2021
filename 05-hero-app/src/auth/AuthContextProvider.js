import { createContext, useReducer, useEffect } from 'react'
import authReducer from './authReducer'

export const AuthContext = createContext()

const init = () => {
  return JSON.parse(localStorage.getItem('user')) || {logged: false};
}

const AuthContextProvider = ({children}) => {
  const [user, dispatch] = useReducer(authReducer, {}, init)

  useEffect(() => {
    localStorage.setItem('user', JSON.stringify(user));
  },[user])

  return (
    <AuthContext.Provider value={{user, dispatch}}>
      {children}
    </AuthContext.Provider>
  )
}

export default AuthContextProvider;
