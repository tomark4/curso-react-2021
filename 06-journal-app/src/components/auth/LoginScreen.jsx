import { useDispatch, useSelector } from 'react-redux';
import { useState } from 'react';
import { Link } from 'react-router-dom';
import { loginWithGoogle, loginWithEmailPassword } from '../../actions/auth';
import validator from 'validator';
import {
    removeError,
    setError
} from '../../actions/ui';

const LoginScreen = () => {
    const dispatch = useDispatch();
    const { ui } = useSelector( state => state );

    const [email, setEmail] = useState("test1@jotta.com");
    const [password, setPassword] = useState("123456");

    const handleSubmit = (e) => {
        e.preventDefault();
        if( !validator.isEmail(email) ){
            dispatch( setError('email invalid') );
            return false;
        }
        if( validator.isEmpty(password) || password.length < 6){
            dispatch( setError('password invalid') );
            return false;
        }
        dispatch( removeError() );
        dispatch( loginWithEmailPassword(email, password) );
    }

    const handleLoginGoogle = () => {
        dispatch( loginWithGoogle() );
    }

    return (
        <>
            <h3 className='auth__title'>Ingresar</h3>
            <form onSubmit={handleSubmit}>
                { ui?.errorMessage &&
                    <div className="auth__alert-error">{ ui?.errorMessage }</div>
                }
                <div>
                    <input
                    type="email"
                    placeholder="Email"
                    name="email"
                    className="auth__input"
                    autoComplete="off"
                    value={email}
                    onChange={(e) => setEmail(e.target.value)}/>
                </div>

                <div>
                    <input
                    type="password"
                    placeholder="Password"
                    name="password"
                    className="auth__input"
                    value={password}
                    onChange={(e) => setPassword(e.target.value)}/>
                </div>

                <div>
                    <button type="submit" className={`btn btn-primary btn-block btn-lg ${ui.loading && 'disabled'}`}
                    disabled={ui.loading}>
                        {ui.loading ? (
                            <>
                                <i class="fa fa-spinner fa-spin"></i> Validando
                            </>
                        ) : 'Ingresar'}
                    </button>
                </div>

                <div className="google-btn" onClick={handleLoginGoogle}>
                    <div className="google-icon-wrapper">
                        <img className="google-icon" src="https://upload.wikimedia.org/wikipedia/commons/5/53/Google_%22G%22_Logo.svg" alt="google button" />
                    </div>
                    <p className="btn-text">
                        <b>Sign in with google</b>
                    </p>
                </div>

            </form>
            <Link to="/auth/register" className="link">Â¿No tienes cuenta? Registrate</Link>
        </>
    )
}

export default LoginScreen
