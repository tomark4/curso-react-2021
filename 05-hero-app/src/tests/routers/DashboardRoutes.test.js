import { mount } from 'enzyme';
import { MemoryRouter } from "react-router";
import { AuthContext } from '../../auth/AuthContextProvider';
import DashboardRoutes from "../../routers/DashboardRoutes";

describe('Pruebas en dashboard routes component', () => {


  test('debe hacer match con el snapshot', () => {
    const wrapper = mount(
      <MemoryRouter>
        <AuthContext.Provider value={{
          user: {logged: true, name:'jose'},
          dispatch: jest.fn()
        }}>
          <DashboardRoutes />
        </AuthContext.Provider>
      </MemoryRouter>
    )
    expect(wrapper).toMatchSnapshot()

    expect( wrapper.find('.text-info').text().trim() ).toBe('Hola, jose')
  })

})
