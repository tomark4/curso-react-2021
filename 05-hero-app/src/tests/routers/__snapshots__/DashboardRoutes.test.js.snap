// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pruebas en dashboard routes component debe hacer match con el snapshot 1`] = `
Array [
  <nav
    className="navbar navbar-expand-sm navbar-dark bg-dark"
  >
    <a
      className="navbar-brand"
      href="/"
      onClick={[Function]}
    >
      Asociaciones
    </a>
    <div
      className="navbar-collapse"
    >
      <div
        className="navbar-nav"
      >
        <a
          aria-current="page"
          className="nav-item nav-link active"
          href="/marvel"
          onClick={[Function]}
          style={Object {}}
        >
          Marvel
        </a>
        <a
          aria-current={null}
          className="nav-item nav-link"
          href="/dc"
          onClick={[Function]}
        >
          DC
        </a>
        <a
          aria-current={null}
          className="nav-item nav-link"
          href="/search"
          onClick={[Function]}
        >
          Search
        </a>
      </div>
    </div>
    <div
      className="navbar-collapse collapse w-100 order-3 dual-collapse2"
    >
      <ul
        className="navbar-nav ml-auto"
      >
        <span
          className="nav-item nav-link text-info"
        >
          Hola, 
          jose
        </span>
        <button
          className="nav-item nav-link btn"
          onClick={[Function]}
        >
          Logout
        </button>
      </ul>
    </div>
  </nav>,
  <div
    className="container mt-3"
  >
    <div
      className="card-columns animate__animated animate__fadeIn"
    >
      <div
        className="card ms-3"
        style={
          Object {
            "maxWidth": "540px",
          }
        }
      >
        <div
          className="row no-gutters"
        >
          <div
            className="col-md-4"
          >
            <img
              alt="Spider Man"
              className="card-img"
              src="./assets/img/heroes/heroes/marvel-spider.jpg"
            />
          </div>
          <div
            className="col-md-8"
          >
            <div
              className="card-body"
            >
              <h5>
                Spider Man
              </h5>
              <p
                className="card-text"
              >
                Peter Parker
              </p>
              <p
                className="text-muted"
              >
                Amazing Fantasy #15
              </p>
              <a
                className="btn btn-outline-success"
                href="/heroe/marvel-spider"
                onClick={[Function]}
              >
                Ver más
              </a>
            </div>
          </div>
        </div>
      </div>
      <div
        className="card ms-3"
        style={
          Object {
            "maxWidth": "540px",
          }
        }
      >
        <div
          className="row no-gutters"
        >
          <div
            className="col-md-4"
          >
            <img
              alt="Captain America"
              className="card-img"
              src="./assets/img/heroes/heroes/marvel-captain.jpg"
            />
          </div>
          <div
            className="col-md-8"
          >
            <div
              className="card-body"
            >
              <h5>
                Captain America
              </h5>
              <p
                className="card-text"
              >
                Steve Rogers
              </p>
              <p
                className="text-muted"
              >
                Captain America Comics #1
              </p>
              <a
                className="btn btn-outline-success"
                href="/heroe/marvel-captain"
                onClick={[Function]}
              >
                Ver más
              </a>
            </div>
          </div>
        </div>
      </div>
      <div
        className="card ms-3"
        style={
          Object {
            "maxWidth": "540px",
          }
        }
      >
        <div
          className="row no-gutters"
        >
          <div
            className="col-md-4"
          >
            <img
              alt="Iron Man"
              className="card-img"
              src="./assets/img/heroes/heroes/marvel-iron.jpg"
            />
          </div>
          <div
            className="col-md-8"
          >
            <div
              className="card-body"
            >
              <h5>
                Iron Man
              </h5>
              <p
                className="card-text"
              >
                Tony Stark
              </p>
              <p
                className="text-muted"
              >
                Tales of Suspense #39
              </p>
              <a
                className="btn btn-outline-success"
                href="/heroe/marvel-iron"
                onClick={[Function]}
              >
                Ver más
              </a>
            </div>
          </div>
        </div>
      </div>
      <div
        className="card ms-3"
        style={
          Object {
            "maxWidth": "540px",
          }
        }
      >
        <div
          className="row no-gutters"
        >
          <div
            className="col-md-4"
          >
            <img
              alt="Thor"
              className="card-img"
              src="./assets/img/heroes/heroes/marvel-thor.jpg"
            />
          </div>
          <div
            className="col-md-8"
          >
            <div
              className="card-body"
            >
              <h5>
                Thor
              </h5>
              <p
                className="card-text"
              >
                Thor Odinson
              </p>
              <p
                className="text-muted"
              >
                Journey into Myster #83
              </p>
              <a
                className="btn btn-outline-success"
                href="/heroe/marvel-thor"
                onClick={[Function]}
              >
                Ver más
              </a>
            </div>
          </div>
        </div>
      </div>
      <div
        className="card ms-3"
        style={
          Object {
            "maxWidth": "540px",
          }
        }
      >
        <div
          className="row no-gutters"
        >
          <div
            className="col-md-4"
          >
            <img
              alt="Hulk"
              className="card-img"
              src="./assets/img/heroes/heroes/marvel-hulk.jpg"
            />
          </div>
          <div
            className="col-md-8"
          >
            <div
              className="card-body"
            >
              <h5>
                Hulk
              </h5>
              <p
                className="card-text"
              >
                Bruce Banner
              </p>
              <p
                className="text-muted"
              >
                The Incredible Hulk #1
              </p>
              <a
                className="btn btn-outline-success"
                href="/heroe/marvel-hulk"
                onClick={[Function]}
              >
                Ver más
              </a>
            </div>
          </div>
        </div>
      </div>
      <div
        className="card ms-3"
        style={
          Object {
            "maxWidth": "540px",
          }
        }
      >
        <div
          className="row no-gutters"
        >
          <div
            className="col-md-4"
          >
            <img
              alt="Wolverine"
              className="card-img"
              src="./assets/img/heroes/heroes/marvel-wolverine.jpg"
            />
          </div>
          <div
            className="col-md-8"
          >
            <div
              className="card-body"
            >
              <h5>
                Wolverine
              </h5>
              <p
                className="card-text"
              >
                James Howlett
              </p>
              <p
                className="text-muted"
              >
                The Incredible Hulk #180
              </p>
              <a
                className="btn btn-outline-success"
                href="/heroe/marvel-wolverine"
                onClick={[Function]}
              >
                Ver más
              </a>
            </div>
          </div>
        </div>
      </div>
      <div
        className="card ms-3"
        style={
          Object {
            "maxWidth": "540px",
          }
        }
      >
        <div
          className="row no-gutters"
        >
          <div
            className="col-md-4"
          >
            <img
              alt="Daredevil"
              className="card-img"
              src="./assets/img/heroes/heroes/marvel-daredevil.jpg"
            />
          </div>
          <div
            className="col-md-8"
          >
            <div
              className="card-body"
            >
              <h5>
                Daredevil
              </h5>
              <p
                className="card-text"
              >
                Matthew Michael Murdock
              </p>
              <p
                className="text-muted"
              >
                Daredevil #1
              </p>
              <a
                className="btn btn-outline-success"
                href="/heroe/marvel-daredevil"
                onClick={[Function]}
              >
                Ver más
              </a>
            </div>
          </div>
        </div>
      </div>
      <div
        className="card ms-3"
        style={
          Object {
            "maxWidth": "540px",
          }
        }
      >
        <div
          className="row no-gutters"
        >
          <div
            className="col-md-4"
          >
            <img
              alt="Hawkeye"
              className="card-img"
              src="./assets/img/heroes/heroes/marvel-hawkeye.jpg"
            />
          </div>
          <div
            className="col-md-8"
          >
            <div
              className="card-body"
            >
              <h5>
                Hawkeye
              </h5>
              <p
                className="card-text"
              >
                Clinton Francis Barton
              </p>
              <p
                className="text-muted"
              >
                Tales of Suspense #57
              </p>
              <a
                className="btn btn-outline-success"
                href="/heroe/marvel-hawkeye"
                onClick={[Function]}
              >
                Ver más
              </a>
            </div>
          </div>
        </div>
      </div>
      <div
        className="card ms-3"
        style={
          Object {
            "maxWidth": "540px",
          }
        }
      >
        <div
          className="row no-gutters"
        >
          <div
            className="col-md-4"
          >
            <img
              alt="Cyclops"
              className="card-img"
              src="./assets/img/heroes/heroes/marvel-cyclops.jpg"
            />
          </div>
          <div
            className="col-md-8"
          >
            <div
              className="card-body"
            >
              <h5>
                Cyclops
              </h5>
              <p
                className="card-text"
              >
                Scott Summers
              </p>
              <p
                className="text-muted"
              >
                X-Men #1
              </p>
              <a
                className="btn btn-outline-success"
                href="/heroe/marvel-cyclops"
                onClick={[Function]}
              >
                Ver más
              </a>
            </div>
          </div>
        </div>
      </div>
      <div
        className="card ms-3"
        style={
          Object {
            "maxWidth": "540px",
          }
        }
      >
        <div
          className="row no-gutters"
        >
          <div
            className="col-md-4"
          >
            <img
              alt="Silver Surfer"
              className="card-img"
              src="./assets/img/heroes/heroes/marvel-silver.jpg"
            />
          </div>
          <div
            className="col-md-8"
          >
            <div
              className="card-body"
            >
              <h5>
                Silver Surfer
              </h5>
              <p
                className="card-text"
              >
                Norrin Radd
              </p>
              <p
                className="text-muted"
              >
                The Fantastic Four #48
              </p>
              <a
                className="btn btn-outline-success"
                href="/heroe/marvel-silver"
                onClick={[Function]}
              >
                Ver más
              </a>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>,
]
`;
